content/en/blog/_posts/2023-01-12-protect-mission-critical-pods-priorityclass/index.md

```diff
@@ -0,0 +1,303 @@
+
+Here's an example. Next, create some environment-specific PriorityClasses:
+
+```yaml
+apiVersion: scheduling.k8s.io/v1
+kind: PriorityClass
+metadata:
+  name: dev-pc
+value: 1000000
+globalDefault: false
+description: >-
+  (Optional) This priority class should only be used for all development pods.
+```
+
+```yaml
+apiVersion: scheduling.k8s.io/v1
+kind: PriorityClass
+metadata:
+  name: preprod-pc
+value: 2000000
+globalDefault: false
+description: >-
+  (Optional) This priority class should only be used for all preprod pods.
+```
+
+```yaml
+apiVersion: scheduling.k8s.io/v1
+kind: PriorityClass
+metadata:
+  name: prod-pc
+value: 4000000
+globalDefault: false
+description: >-
+  (Optional) This priority class should only be used for all prod pods.
+```
+
+Use `kubectl create -f <FILE.YAML>` command to create a pc and `kubectl get pc` to check its status.
+
+```console
+$ kubectl get pc
+NAME                      VALUE        GLOBAL-DEFAULT   AGE
+dev-pc                    1000000      false            3m13s
+preprod-pc                2000000      false            2m3s
+prod-pc                   4000000      false            7s
+system-cluster-critical   2000000000   false            82m
+system-node-critical      2000001000   false            82m
+```
+
+The new PriorityClasses are in place now. A small change is needed in the pod manifest or pod template (in a ReplicaSet or Deployment). In other words, we need to specify the priority class name at `.spec.priorityClassName` (which is a string value).
+
+First update the previous production pod manifest file to have a PriorityClass assigned, then delete the Production pod and recreate it. You can't edit the priority class for a Pod that already exists.
+
+That change seems successful; the status of pods has been updated:
+
+```console
+$ kubectl get pods --show-labels
+NAME            READY   STATUS    	RESTARTS   AGE   LABELS
+dev-nginx       1/1     Terminating	0          55s   env=dev
+preprod-nginx   1/1     Running   	0          55s   env=preprod
+prod-nginx      0/1     Pending   	0          55s   env=prod
+```
+
+The dev-nginx pod is getting terminated. Once that is successfully terminated and there are enough resources for the prod pod, the control plane can schedule the prod pod:
+
+```console
+Warning   FailedScheduling   pod/prod-nginx    0/2 nodes are available: 1 Insufficient cpu, 2 Insufficient memory.
+Normal    Preempted          pod/dev-nginx     by default/prod-nginx on node node01
+Normal    Killing            pod/dev-nginx     Stopping container dev-nginx
+Normal    Scheduled          pod/prod-nginx    Successfully assigned default/prod-nginx to node01
+Normal    Pulling            pod/prod-nginx    Pulling image "nginx"
+Normal    Pulled             pod/prod-nginx    Successfully pulled image "nginx"
+Normal    Created            pod/prod-nginx    Created container prod-nginx
+Normal    Started            pod/prod-nginx    Started container prod-nginx
+```
+
```

Comment:
Readers might assume that these PriorityClasses are somehow enforced, which isn't true.

You could add something similar to:
```suggestion

## Enforcement

When you set up PriorityClasses, they exist just how you defined them. However, people
(and tools) that make changes to your cluster are free to set any PriorityClass, or to not
set any PriorityClass at all.
However, you can use other Kubernetes features to make sure that the priorities you wanted
are actually applied.

As an alpha feature, you can define a [ValidatingAdmissionPolicy](/blog/2022/12/20/validating-admission-policies-alpha/) and a ValidatingAdmissionPolicyBinding so that, for example,
Pods that go into the `prod` namespace must use the `prod-pc` PriorityClass.
With another ValidatingAdmissionPolicyBinding you ensure that the `preprod` namespace
uses the `preprod-pc` PriorityClass, and so on.
In *any* cluster, you can enforce similar controls using external projects such as
[Kyverno](https://kyverno.io/) or [Gatekeeper](https://open-policy-agent.github.io/gatekeeper/),
through validating admission webhooks.

However you do it, Kubernetes gives you options to make sure that the PriorityClasses are
used how you wanted them to be, or perhaps just to
[warn](https://open-policy-agent.github.io/gatekeeper/website/docs/violations/#warn-enforcement-action)
users when they pick an unsuitable option.

```

```

Reply:
This is a nice piece of information, though I haven't tried it yet, I will surely give it a try.

Since this feature is in alpha, Maybe in the future we can create a part-2 on this and can link that with this article. What do you think? 

Also about the enforcement part, yes true we can add below section.

## Enforcement
When you set up PriorityClasses, they exist just how you defined them. However, people (and tools) that make changes to your cluster are free to set any PriorityClass, or not to set any PriorityClass at all.

Reply:
ValidatingAdmissionPolicy is alpha, but [admission webhooks](https://kubernetes.io/docs/reference/access-authn-authz/extensible-admission-controllers/#what-are-admission-webhooks) are stable. It's important to cover the overall admission control mechanism at least in outline, because it is relevant.

If you want to explain that aspect in another article, turning this into 2 parts, that's absolutely fine -  but I'd want to see a PR opened and drafted for part 2 before we merge the PR for part 1.

If you do opt to add a second part, that gives you room to cover some of the details from https://kubernetes.io/docs/concepts/policy/resource-quotas/#resource-quota-per-priorityclass

---

The minimum here is to mention that admission control is often an important part of making priorityclasses work right; you must mention that admission control concept somehow. Maybe in only 50 more words.

If you want to go beyond that, you have options. I'm fine with making this a 2 part piece, even if that means a delay. It's your article, so you do get to choose which option you prefer.

Reply:
That makes sense, I will go with your suggestion. 

But still maybe in near future, I will write an article about ValidatingAdmissionPolicy and resource-quota-per-priorityclass for sure.

Thank you for all your suggestions and time.